### ISSUE TITLE
Stack Monitoring Templates are not installed until rolling upgrade is complete

### ISSUE THREAD

**Issue by jbaiera** [elastic/elasticsearch#85247]
The Monitoring plugin currently installs templates for stack monitoring data streams in the format of `.monitoring-{product}-8-mb`. These templates are required for regular operation of Metricbeat when used for monitoring Elasticsearch.

These templates are installed on the currently elected master node as part of the `MonitoringTemplateRegistry` operation. When performing a rolling upgrade, it is advised that all data nodes are upgraded to the new versions first, followed by the master nodes. The rolling upgrade steps on cloud follow this pattern. 

During a rolling upgrade, nodes are started on the newer version of the stack. These nodes include the newer version of Metricbeat used for monitoring. While the rolling upgrade continues, these updated Metricbeat instances begin shipping data to the cluster using the new `.monitoring-es-8-mb` format. Since the master has not been upgraded yet, there are no templates installed to handle the incoming data. The monitoring data is incorrectly stored in a dynamically allocated index instead of a data stream. Furthermore, the templates include alias fields that would correctly map the new monitoring data model to the old storage format. Since the templates are not present, the mappings are not applied correctly to the monitoring index. This breaks the stack monitoring UI. 

The fix is to wait for the templates to be installed after the rolling upgrade and to delete the erroneously created `.monitoring-es-8-mb` index. If this is not done after a few days, the CleanerService configured in the Monitoring plugin will clear the index itself, and the data stream will be created correctly.

One option for fixing this issue is that Metricbeat should hold all outbound monitoring data until the monitoring templates are present on the cluster being written to. This keeps the application from breaking, but a rolling upgrade can take a while to complete depending on how large the cluster is. A small 5 node deployment took 15 minutes to reach a point where the elected master node was on the newer version, during which all other data nodes would be blocked from sending monitoring data.

Discussion points:
* Is the window of time during a rolling upgrade worth preserving monitoring data through? Do we have an idea of how long this normally takes? The larger the cluster is and longer the window lasts, the more important it seems to make sure we're grabbing data.
* Should we find a way to land these templates earlier? They are configured to install on master because that is how Monitoring has historically worked. Templates are used as a signal to indicate readiness of the legacy parts of the monitoring plugin. We would need to do some analysis to see if anything breaks if they are installed from any other node.

**Comment by jbaiera** [comment]
To address the second bullet point:

We install the stack monitoring templates with the same registry that we install the legacy templates with. We could split this registry apart (probably a good idea anyway) into two registries so that stack monitoring templates have the possibility to be installed sooner than the legacy ones.

This is still an issue if we introduce new functionality over time though. If we add features to the templates that require master to be updated before they are applied (for instance, new settings on templates for TSDB) then we need to once again wait for the master node to be upgraded before installing the templates.

**Comment by matschaffer** [comment]
> The larger the cluster is and longer the window lasts, the more important it seems to make sure we're grabbing data.

++ to this. A long running upgrade is a sensitive time for a cluster. We want as much operational telemetry as we can get incase of issues encountered during the upgrade.

**Comment by jbaiera** [comment]
We discussed this a while back but the challenges with fixing this in Elasticsearch are likely hard to impossible to overcome. If new mappings rely on features that are not yet available until the cluster is updated, then there is no way to install content. If we were to install content that is backwards compatible we would still need to maintain the way to install it when it is incompatible.

The easier fix here is to upgrade clients that need the new features in Elasticsearch after Elasticsearch is fully upgraded. If incompatible changes are present, then the clients will need to continue writing in the old way until they can detect that Elasticsearch is able to accept the new data format.

It's possible that we may rework how these resources are installed in the future, the current way that they are installed is not the most scaleable or elegant way of doing things, but until that happens I don't think there's a whole lot that can be done about it. Closing for now.
